src/client/Microsoft.Identity.Client/Internal/Broker/BrokerSilentRequest.cs

```diff
@@ -0,0 +1,95 @@
+            _silentParameters = acquireTokenSilentParameters;
+            _serviceBundle = serviceBundle;
+            Broker = broker;
+        }
+
+        public async Task<MsalTokenResponse> SendTokenRequestToBrokerAsync()
+        {
+            //TODO: call can invoke broker here
+            _authenticationRequestParameters.RequestContext.Logger.Info(LogMessages.CanInvokeBrokerAcquireTokenWithBroker);
+
+            return await SendAndVerifyResponseAsync().ConfigureAwait(false);
+        }
+
+        private async Task<MsalTokenResponse> SendAndVerifyResponseAsync()
+        {
+            CreateRequestParametersForBroker();
+
+            MsalTokenResponse msalTokenResponse =
+                await Broker.AcquireTokenUsingBrokerAsync(_brokerPayload).ConfigureAwait(false);
+
+            ValidateResponseFromBroker(msalTokenResponse);
+            return msalTokenResponse;
+        }
+
+        internal void CreateRequestParametersForBroker()
+        {
+            _brokerPayload.Add(BrokerParameter.Authority, _authenticationRequestParameters.Authority.AuthorityInfo.CanonicalAuthority);
+            string scopes = EnumerableExtensions.AsSingleString(_authenticationRequestParameters.Scope);
+
+            _brokerPayload.Add(BrokerParameter.Scope, scopes);
+            _brokerPayload.Add(BrokerParameter.ClientId, _authenticationRequestParameters.ClientId);
+            _brokerPayload.Add(BrokerParameter.CorrelationId, _authenticationRequestParameters.RequestContext.Logger.CorrelationId.ToString());
+            _brokerPayload.Add(BrokerParameter.ClientVersion, MsalIdHelper.GetMsalVersion());
+            _brokerPayload.Add(BrokerParameter.RedirectUri, _serviceBundle.Config.RedirectUri);
+            string extraQP = string.Join("&", _authenticationRequestParameters.ExtraQueryParameters.Select(x => x.Key + "=" + x.Value));
+            _brokerPayload.Add(BrokerParameter.ExtraQp, extraQP);
+            _brokerPayload.Add(BrokerParameter.ExtraOidcScopes, BrokerParameter.OidcScopesValue);
+            _brokerPayload.Add(BrokerParameter.LoginHint, _silentParameters.LoginHint);
+#pragma warning disable CA1305 // Specify IFormatProvider
+            _brokerPayload.Add(BrokerParameter.ForceRefresh, _silentParameters.ForceRefresh.ToString());
+#pragma warning restore CA1305 // Specify IFormatProvider
+        }
+
+        internal void ValidateResponseFromBroker(MsalTokenResponse msalTokenResponse)
+        {
+            _authenticationRequestParameters.RequestContext.Logger.Info(LogMessages.CheckMsalTokenResponseReturnedFromBroker);
+            if (msalTokenResponse.AccessToken != null)
+            {
+                _authenticationRequestParameters.RequestContext.Logger.Info(
+                    LogMessages.BrokerResponseContainsAccessToken +
+                    msalTokenResponse.AccessToken.Count());
+                return;
+            }
+            if (msalTokenResponse.Error != null)
```

Comment:
Shouldn't you check for an error before checking for an access token? Just in case.

Reply:
yes